File: Algorithms/138.CopyListWithRandomPointer.cpp
class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
        Clone(head);
        CopyRandomPointer(head);
        return restore(head);
    }
    void Clone(RandomListNode *head) {
        while (head) {
            RandomListNode *clonedNode = new RandomListNode(head->label);
            clonedNode->next = head->next;
            head->next = clonedNode;
            head = clonedNode->next;
        }
    }
    void CopyRandomPointer(RandomListNode *head) {
        while (head) {
            RandomListNode *clonedNode = head->next;
            if (head->random) {
                clonedNode->random = head->random->next;
            }
            head = clonedNode->next;
        }
    }
    RandomListNode *restore(RandomListNode *head) {
        RandomListNode *pClonedHead = new RandomListNode(-1);
        RandomListNode *pClonedNode = pClonedHead;
        while (head) {
            pClonedNode->next = head->next;
            pClonedNode = pClonedNode->next;
            head->next = pClonedNode->next;
            head = head->next;
        }
        return pClonedHead->next;
    }
};

File: Algorithms/138.CopyListWithRandomPointer.py
class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):
        self.Clone(head)
        self.CopyRandomPointer(head)
        return self.restore(head)

    def Clone(self, head):
        while head:
            clonedNode = RandomListNode(head.label)
            clonedNode.next = head.next
            head.next = clonedNode
            head = clonedNode.next

    def CopyRandomPointer(self, head):
        while head:
            clonedNode = head.next
            if head.random:
                clonedNode.random = head.random.next
            head = clonedNode.next

    def restore(self, head):
        pClonedHead = RandomListNode(-1)
        pClonedNode = pClonedHead
        while head:
            pClonedNode.next = head.next
            pClonedNode = pClonedNode.next
            head.next = pClonedNode.next
            head = head.next
        return pClonedHead.next

